vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,361 @@
+      {
+        std::vector<VW::string_view> pair = split(ns_values[i], ":");
+        std::string ns = " ";
+        if (pair.size() != 2 || pair[1].empty() || !check_if_float(pair[1]))
+        { THROW("Malformed namespace value pair: " << ns_values[i]); }
+        else if (!pair[0].empty())
+        {
+          ns = {pair[0].begin(), pair[0].end()};
+        }
+        size_t end_idx = pair[1].length();
+        ns_value.insert(std::make_pair(ns, parseFloat(pair[1].data(), end_idx)));
+      }
+    }
+
+    // If no header present, will use empty features
+    if (all->csv_no_header && _header_fn.empty())
+    {
+      for (size_t i = 0; i < elements.size(); i++)
+      {
+        _header_fn.emplace_back(std::string());
+        _header_ns.emplace_back(std::string());
+      }
+    }
+
+    if (!_header_fn.empty() && elements.size() != _header_fn.size())
+    {
+      THROW(
+          "CSV line has " << elements.size() << " elements, but the header has " << _header_fn.size() << " elements!");
+    }
+    else if (!all->csv_no_header && _header_fn.empty())
+    {
+      for (size_t i = 0; i < elements.size(); i++)
+      {
+        if (all->csv_remove_quotes) { remove_quotation_marks(elements[i]); }
+
+        // Seperate the feature name and namespace from the header.
+        std::vector<VW::string_view> splitted = split(elements[i], all->csv_ns_separator);
+        VW::string_view feature_name;
+        VW::string_view ns;
+        if (splitted.size() == 1) { feature_name = elements[i]; }
+        else if (splitted.size() == 2)
+        {
+          ns = splitted[0];
+          feature_name = splitted[1];
+        }
+        else
+        {
+          THROW("Malformed header for feature name and namespace separator: " << elements[i]);
+        }
+        _header_fn.emplace_back(feature_name);
+        _header_ns.emplace_back(ns);
+      }
+    }
+    else
+    {
+      parse_example(all, ae, elements);
+    }
+  }
+}
+
+void parser::parse_example(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  if (label_list.empty() && !all->csv_label.empty())
+  {
+    VW::string_view csv_label(all->csv_label);
+    label_list = list_handler(csv_label, csv_line.size(), "label");
+  }
+  if (tag_list.empty() && !all->csv_tag.empty())
+  {
+    VW::string_view csv_tag(all->csv_tag);
+    tag_list = list_handler(csv_tag, csv_line.size(), "tag");
+    if (tag_list.size() > 1) { THROW("Can only have one tag at most per example!"); }
+  }
+
+  if (!label_list.empty()) { parse_label(all, ae, csv_line); }
+  if (!tag_list.empty()) { parse_tag(all, ae, csv_line); }
+
+  parse_namespaces(all, ae, csv_line);
+}
+
+void parser::parse_label(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  all->example_parser->lbl_parser.default_label(ae->l);
+
+  std::string label_content;
+  char label_part_separator = ' ';
+  if (all->csv_multilabels) { label_part_separator = ','; }
```

Comment:
Could you explain the `csv_multilabels` special case?

Reply:
For labels, so far I'm just convert it into the vw format one for it to work https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Input-format#labels

So e.g. For [Iris3.csv](https://github.com/HollowMan6/vowpal_wabbit/files/8754903/Iris3.csv)

Assume it as a multilabels example, unlike the simple labels, we have to combine the two/more columns (label components) with `,` instead of spaces by specifying `the --csv_multilabels`.
`vw --csv Iris3.csv --csv_label -3,-2 --csv_tag -1 --csv_multilabels --multilabel_oaa 5`
![image](https://user-images.githubusercontent.com/43995067/169830583-647ec987-c5aa-404a-b4fa-75f36540210c.png)

Reply:
I am assuming you built this because the label itself contains commas so it's ambiguous with this scheme.  This is tricky because each line is not guaranteed to have the same number of labels for a multilabel example.

Interesting, there are certainly other labels which also contain commas, so this will be a recurring problem. Perhaps this is reason for us to properly support quotes? Maybe we simply specify that newlines in quotes are not allowed to keep the line splitting simple still?

Reply:
Okay sounds good. Thanks for the discussion!